src/Interpreters/GraceHashJoin.cpp

```diff
@@ -356,52 +353,66 @@ bool GraceHashJoin::hasMemoryOverflow(const InMemoryJoinPtr & hash_join_) const
     return hasMemoryOverflow(total_rows, total_bytes);
 }
 
-GraceHashJoin::Buckets GraceHashJoin::rehashBuckets(size_t to_size)
+GraceHashJoin::Buckets GraceHashJoin::rehashBuckets()
 {
     std::unique_lock lock(rehash_mutex);
-    size_t current_size = buckets.size();
 
-    if (to_size <= current_size)
-        return buckets;
+    if (!isPowerOf2(buckets.size())) [[unlikely]]
+        throw Exception(ErrorCodes::LOGICAL_ERROR, "Number of buckets should be power of 2 but it's {}", buckets.size());
 
-    chassert(isPowerOf2(to_size));
+    const size_t to_size = buckets.size() * 2;
+    size_t current_size = buckets.size();
 
     if (to_size > max_num_buckets)
     {
-        throw Exception(ErrorCodes::LIMIT_EXCEEDED,
+        throw Exception(
+            ErrorCodes::LIMIT_EXCEEDED,
             "Too many grace hash join buckets ({} > {}), "
             "consider increasing grace_hash_join_max_buckets or max_rows_in_join/max_bytes_in_join",
-            to_size, max_num_buckets);
+            to_size,
+            max_num_buckets);
     }
 
     LOG_TRACE(log, "Rehashing from {} to {}", current_size, to_size);
 
-    buckets.reserve(to_size);
-    for (size_t i = current_size; i < to_size; ++i)
-        addBucket(buckets);
+    addBuckets(to_size - current_size);
 
     return buckets;
 }
 
-void GraceHashJoin::addBucket(Buckets & destination)
+void GraceHashJoin::addBuckets(const size_t bucket_count)
 {
-    // There could be exceptions from createStream, In ci tests
-    // there is a certain probability of failure in allocating memory, see memory_tracker_fault_probability.
-    // It may terminate this thread and leave a broken hash_join, and another thread cores when it tries to
-    // use the broken hash_join. So we print an exception message here to help debug.
-    try
-    {
-        auto & left_file = tmp_data->createStream(left_sample_block);
-        auto & right_file = tmp_data->createStream(prepareRightBlock(right_sample_block));
+    // Exception can be thrown in number of cases:
```

Comment:
If an exception is thrown, won't the query be cancelled? Do we see an inconsistent state of `buckets` because of another thread accessing it when the exception has already been thrown (and the mutex released), but before the query was cancelled?

Reply:
Yes, that's what I see if intentionally add exception in `addBuckets()` while creating new buckets

Reply:
Expanded part of `02273_full_sort_join.sql.j2` for grace join:
https://pastila.nl/?00b6056d/cde7c624db96e692a69ef05cc0132f95

Traces for grace hash join with an intentionally added exception when trying to create 5th bucket:
https://pastila.nl/?022538cf/c96e4cb05f99476123270f2ccad3fcdf

Intentionally added exception in log:
```
... <Error> GraceHashJoin: Can't create bucket 4 due to error: Code: 290. DB::Exception: Too many buckets. (LIMIT_EXCEEDED) (version 23.7.1.1)
2
````

Reply:
The exception is catch and thrown in `PipelineExecutor` asynchronously. I think before the exception is really throw out, other threads still have chance to run and trigger this exception.